homeassistant/components/asuswrt/diagnostics.py

```diff
@@ -0,0 +1,82 @@
+"""Diagnostics support for Asuswrt."""
+from __future__ import annotations
+
+from typing import Any
+
+from homeassistant.components.diagnostics import async_redact_data
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import device_registry as dr, entity_registry as er
+
+from .const import DATA_ASUSWRT, DOMAIN
+from .router import AsusWrtRouter
+
+TO_REDACT = {CONF_PASSWORD, CONF_USERNAME}
+
+
+async def async_get_config_entry_diagnostics(
+    hass: HomeAssistant, entry: ConfigEntry
+) -> dict[str, dict[str, Any]]:
+    """Return diagnostics for a config entry."""
+    data = {"entry": async_redact_data(entry.as_dict(), TO_REDACT)}
+
+    router: AsusWrtRouter = hass.data[DOMAIN][entry.entry_id][DATA_ASUSWRT]
+
+    # Gather information how this AsusWrt device is represented in Home Assistant
+    device_registry = dr.async_get(hass)
+    entity_registry = er.async_get(hass)
+    hass_device = device_registry.async_get_device(
+        identifiers=router.device_info["identifiers"]
+    )
+    if not hass_device:
+        return data
+
+    data["device"] = {
+        "name": hass_device.name,
+        "name_by_user": hass_device.name_by_user,
+        "disabled": hass_device.disabled,
+        "disabled_by": hass_device.disabled_by,
+        "device_info": async_redact_data(dict(router.device_info), {"identifiers"}),
+        "entities": [],
+        "tracked_devices": [],
+    }
+
+    hass_entities = er.async_entries_for_device(
+        entity_registry,
+        device_id=hass_device.id,
+        include_disabled_entities=True,
+    )
+
+    for entity_entry in hass_entities:
+        state = hass.states.get(entity_entry.entity_id)
+        state_dict = None
+        if state:
+            state_dict = state.as_dict()
+            # The context doesn't provide useful information in this case.
+            state_dict.pop("context", None)
```

Comment:
There is a bug here. This structure should be copied because the dict is saved in a state variable and the `pop` is going to delete the internal data.
This issue is for sure present also in Tuya diagnostic because I started from there.

Reply:
The code makes new dicts:

- [`Context.as_dict`](https://github.com/home-assistant/core/blob/dev/homeassistant/core.py#L704-L706)
- [`State.as_dict`](https://github.com/home-assistant/core/blob/dev/homeassistant/core.py#L1066-L1088)

Reply:
Only first time, but it store state in [self._state](https://github.com/home-assistant/core/blob/8245ff7473a4242faad007a53f610d9dc98b46b0/homeassistant/core.py#L1074). So next time that `state.as_dict` is called, it re-use `self._state` that not contain anymore context (because was pop).

Reply:
oh snap!

Reply:
Fixed in #65693